cluster-controller/src/main/java/io/strimzi/controller/cluster/model/AbstractModel.java

```diff
@@ -554,4 +563,71 @@ protected static EnvVar buildEnvVar(String name, String value) {
                 // On duplicates, last in wins
                 (u, v) -> v));
     }
+
+    protected ResourceRequirements resources() {
+        if (resources != null) {
+            ResourceRequirementsBuilder builder = new ResourceRequirementsBuilder();
+            Resources.CpuMemory limits = resources.getLimits();
+            if (limits != null
+                    && limits.getMilliCpu() > 0) {
+                builder.addToLimits("cpu", new Quantity(limits.getCpuFormatted()));
+            }
+            if (limits != null
+                    && limits.getMemory() > 0) {
+                builder.addToLimits("memory", new Quantity(limits.getMemoryFormatted()));
+            }
+            Resources.CpuMemory requests = resources.getRequests();
+            if (requests != null
+                    && requests.getMilliCpu() > 0) {
+                builder.addToRequests("cpu", new Quantity(requests.getCpuFormatted()));
+            }
+            if (requests != null
+                    && requests.getMemory() > 0) {
+                builder.addToRequests("memory", new Quantity(requests.getMemoryFormatted()));
+            }
+            return builder.build();
+        }
+        return null;
+    }
+
+    public void setResources(Resources resources) {
+        this.resources = resources;
+    }
+
+    public void setJvmOptions(JvmOptions jvmOptions) {
+        this.jvmOptions = jvmOptions;
+    }
+
+    protected String javaHeapOptions(long maxNeeded, double fraction) {
+        StringBuilder result = new StringBuilder();
+        String xms = jvmOptions != null ? jvmOptions.getXms() : null;
+
+        if (xms != null) {
+            result.append("-Xms").append(xms).append(' ');
+        }
+
+        String xmx = jvmOptions != null ? jvmOptions.getXmx() : null;
+        if (xmx != null) {
+            // Honour explicit max heap
+            result.append("-Xmx").append(xmx).append(' ');
+        } else if (resources != null
+                && resources.getRequests() != null
+                && resources.getRequests().getMemory() > 0) {
+            long avail = resources.getRequests().getMemory();
```

Comment:
This will be a problem in environments like OpenShift Online Pro & Starter. Those environments completely ignore any memory requests. Instead, the requested memory is set automatically based on a percentage of the memory limit. If you want to do this kind of logic, you'll have to do it from inside the container based on reading the various CGroups values actually given to this container. Or, perform this logic based off the limit and not the request. Or, perhaps document that users of Online Pro & Starter need to always set a request AND a limit. But it would be better to avoid having to document that special case.

Reply:
The problem with using the limit is we can't be sure we'll actually get that much memory. If we get less then it screws up the calculation wrt leaving memory unused by the JVM for use as page cache. The idea behind using the request was it meant that calculation wasn't bogus (even if the OS never actually used the 'spare' memory for caching pages of the kafka log).

Reply:
I get that. I'm just pointing out that in some OpenShift environments this will bite users because the request this calculation is based off of is completely ignored and overwritten by OpenShift.